File: DATA_STRUCTURES_PY/Queue___(based_list).py
# /*Queue FIFO - очередь на основе списка*/
class Queue:                        # Класс описывающий очередь
    def __init__(self, data): pass  # Конструктор
    def push(self, elm): pass       # Добавить элемент в конец очереди
    def pop(self): pass             # Удалить первый элемент
    def show(self): pass            # Печать
    def empty(self): pass           # true если пуста, иначе false
    def size(self): pass            # Кол-во элементов
    def back(self): pass            # Последний элемент в очереди
    def front(self): pass           # Первый элемент
"""-------------------------------------------"""

class Queue:
    def __init__(self, data=None):
        if data is None:
            self._size = 0
            self._node = []
        else:
            self._size = 1
            self._node = [data]

    def push(self, elm):
        self._node.append(elm)
        self._size += 1

    def pop(self):
        if self._size == 0:
            return
        self._node.pop()
        self._size -= 1

    def show(self):
        if self._size == 0:
            return
        for i in range(self._size):
            print(f"({self._node[i]})", end="<-")
        print('')

    def empty(self):
        return self._size == 0

    def size(self):
        return self._size

    def back(self):
        return self._node[self._size - 1] if self._size >= 1 else None

    def front(self):
        return self._node[0] if self._size > 0 else None


def main():
    que = Queue()
    for i in range(1, 16):
        que.push(i)
    que.show()
    for i in range(5):
        que.pop()
    que.show()

if __name__ == "__main__":
    main()

File: DATA_STRUCTURES_CPP/Queue___(based_list).cpp
/*Queue FIFO - очередь на основе списка*/
#include <iostream>
#include <list>
#include <queue>

using std::list;
using std::queue;
using std::cout;


template <typename Type>
class Queue {							// Класс описывающий очередь
	list<Type> node;					// Класс описывающий узел очереди
	size_t cnt;							// Кол-во элементов
public:
	Queue();							// Конструктор
	Queue(const Type &elm);				// Конструктор с параметром
	Queue(const Queue<Type> &obj);		// Конструктор копирования
	~Queue();							// Деструктор
	void push(const Type &elm);			// Добавить в конец
	void pop();							// Вытащить первый элемент из очереди
	void show();						// Вывод на экран
	void empty() {return cnt == 0;}		// true если пуста, иначе false
	void size()  {return cnt;}			// Кол-во элементов
	const Type& back()  const { return node.back(); }	// Последний элемент в очереди
	const Type& front() const { return node.front(); }  // Первый элемент в очереди
};


template <typename Type>
Queue<Type>::Queue() : cnt(0) {};

template <typename Type>
Queue<Type>::Queue(const Type& elm) {
	cnt = 1;
	node.push_back(elm);
}

template <typename Type>
Queue<Type>::Queue(const Queue<Type>& obj) {
	if (obj.node.size() == 0) {
		this->Queue<Type>::Queue();
		return;
	}
	for (auto iter = obj.node.begin(); iter != obj.node.end(); ++iter)
		push(*iter);
}

template <typename Type>
Queue<Type>::~Queue() {
	node.clear();
	cnt = 0;
}

template <typename Type>
void Queue<Type>::push(const Type &elm) {
	node.push_back(elm);
	cnt++;
}

template <typename Type>
void Queue<Type>::pop() {
	if (node.size() == 0)
		return;
	node.pop_back();
	cnt--;
}

template <typename Type>
void Queue<Type>::show() {
	if (cnt == 0)
		return;
	for (auto iter = node.begin(); iter != node.end(); ++iter)
		cout << "-------";
	cout << '\n';
	for (auto iter = node.begin(); iter != node.end(); ++iter)
		cout << '(' << * iter << ")<-";
	cout << '\n';
	for (auto iter = node.begin(); iter != node.end(); ++iter)
		cout << "-------";
	cout << '\n';
}


int main() {
	Queue<int> mq;
	for (int i = 1; i <= 15; ++i)
		mq.push(i);
	mq.show();
	for (int i = 0; i < 5; ++i)
		mq.pop();
	mq.show();
}